File: alternative/medium/coin_change.py
from typing import List

def coinChange(coins: List[int], amount: int) -> int:
    dp = [amount + 1] * (amount + 1)
    dp[0] = 0

    for a in range(1, amount + 1):
        for c in coins:
            if c <= a:
                dp[a] = min(dp[a], dp[a - c] + 1)
    
    return -1 if dp[amount] > amount else dp[amount]


assert coinChange([1,2,5], 11) == 3
assert coinChange([2], 3) == -1
assert coinChange([1], 0) == 0

File: src/medium/coin_change.rs
#![allow(dead_code)]
pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
    let mut dp = vec![amount + 1; (amount + 1) as usize];
    dp[0] = 0;

    for a in 1..=amount {
        for &c in &coins {
            if c <= a {
                dp[a as usize] = dp[a as usize].min(dp[(a - c) as usize] + 1);
            }
        }
    }

    if dp[amount as usize] > amount {
        -1
    } else {
        dp[amount as usize]
    }
}

#[test]
fn test_coin_change() {
    assert_eq!(coin_change(vec![1, 2, 5], 11), 3);
    assert_eq!(coin_change(vec![2], 3), -1);
    assert_eq!(coin_change(vec![1], 0), 0);
    assert_eq!(coin_change(vec![1], 1), 1);
    assert_eq!(coin_change(vec![1], 2), 2);
}
